### ISSUE TITLE
Oracle sqlldr fails on large CLOB fields

### ISSUE THREAD

**Issue by dvdrndlph** [slingdata-io/sling-cli#446]
## Issue Description
The .ctl file generated by Sling limits the size of source/target CLOB columns to 400,000 characters.

- Description of the issue: 
If source column has over 400K characters, sqlldr throws an error like the following:
```
Record 54969: Rejected - Error on table "CABOODLE"."ClinicalNoteTextFact_TMP7J", column "Text".
Field in data file exceeds maximum length
```
If I set SLING_SAVE_TEMP to true, tweak the .ctl file to accept 900K characters, then rerun sqlldr, the job succeeds. 

To make this work with CLOBs of arbitrary length would be ideal, but this would take a much bigger change to your sqlldr configuration. 

An interim idea would be to support an input parameter and/or environment variable to make the CLOB size configurable, e.g., "SLING_MAX_CLOB_SIZE=9999999."

Otherwise, a larger default setting (of perhaps 1,000,000) would be helpful. 

- Sling version (`sling --version`): 
1.2.22

- Operating System (`linux`, `mac`, `windows`): 
Windows

**Comment by flarco** [comment]
Yeah I think upping the default will work. Could you paste/summarize the
changes you made to your CTL file?

**Comment by dvdrndlph** [comment]
Sorry, my .ctl file got deleted. This should be as easy as searching for "400000" in your code base and changing it to some bigger number.

I think you just need to change this file: `core/dbio/database/database_oracle.go`.

It would be nice to have an environment variable to be able to work around this issue without needing a code change. But expediency is good.

**Comment by flarco** [comment]
Did you mean 4000? I was just looking, and could find anything.
Found this: https://stackoverflow.com/questions/46090485/sql-loader-load-string-greater-than-4000-characters-using-sql-loader

**Comment by dvdrndlph** [comment]
No, I see 400000 on line 424 of `core/dbio/database/database_oracle.go`.

**Comment by flarco** [comment]
Thanks. I don't remember committing that. 😄 
Will change soon.

**Comment by dvdrndlph** [comment]
Here is the output from my successful run:
```
SQL*Loader: Release 19.0.0.0.0 - Production on Thu Nov 21 01:40:33 2024
Version 19.3.0.0.0

Copyright (c) 1982, 2019, Oracle and/or its affiliates.  All rights reserved.

Control File:   C:/cygwin64/tmp/oracle.caboodle.clinicalnotetextfact_tmp7j.sqlldr.1732172319951.kRf.ctl
Data File:      C:/cygwin64/tmp/oracle.caboodle.clinicalnotetextfact_tmp7j.1732172319953.MAs.temp.csv
  Bad File:     C:/cygwin64/tmp/bad.log
  Discard File: C:/cygwin64/tmp/oracle.caboodle.clinicalnotetextfact_tmp7j.1732172319953.MAs.temp.dsc 
 (Allow 0 discards)

Number to load: ALL
Number to skip: 1
Errors allowed: 0
Continuation:    none specified
Path used:      Direct - with parallel option.

Table "CABOODLE"."ClinicalNoteTextFact_TMP7J", loaded from every logical record.
Insert option in effect for this table: APPEND
TRAILING NULLCOLS option in effect

   Column Name                  Position   Len  Term Encl Datatype
------------------------------ ---------- ----- ---- ---- ---------------------
"ClinicalNoteTextKey"               FIRST     *   ,  O(") CHARACTER            
"ClinicalNoteKey"                    NEXT     *   ,  O(") CHARACTER            
"Text"                               NEXT *****   ,  O(") CHARACTER            
    Maximum field length is 900000
    NULL if "Text" = BLANKS
"Count"                              NEXT     *   ,  O(") CHARACTER            
"_CreationInstant"                   NEXT     *   ,  O(") CHARACTER            
    SQL string for column : "TO_TIMESTAMP(:"_CreationInstant", 'YYYY-MM-DD HH24:MI:SS.FF6')"
"_LastUpdatedInstant"                NEXT     *   ,  O(") CHARACTER            
    SQL string for column : "TO_TIMESTAMP(:"_LastUpdatedInstant", 'YYYY-MM-DD HH24:MI:SS.FF6')"
"_IsInferred"                        NEXT     *   ,  O(") CHARACTER            
"_IsDeleted"                         NEXT     *   ,  O(") CHARACTER            
"_PrimaryPackageToImpactRecord"       NEXT *****   ,  O(") CHARACTER            
    Maximum field length is 400000
    NULL if "_PrimaryPackageToImpactRecord" = BLANKS
"_MostRecentPackageToImpactRecord"       NEXT *****   ,  O(") CHARACTER            
    Maximum field length is 400000
    NULL if "_MostRecentPackageToImpactRecord" = BLANKS
"_NumberOfSources"                   NEXT     *   ,  O(") CHARACTER            
"_HasSourceClarity"                  NEXT     *   ,  O(") CHARACTER            
"_DeletedFromSourceClarity"          NEXT     *   ,  O(") CHARACTER            

SQL*Loader-281: Warning: ROWS parameter ignored in parallel mode.


Table "CABOODLE"."ClinicalNoteTextFact_TMP7J":
  100000 Rows successfully loaded.
  0 Rows not loaded due to data errors.
  0 Rows not loaded because all WHEN clauses were failed.
  0 Rows not loaded because all fields were null.
```
The .ctl file can control how much data is allowed for a character data field. And I am quite sure this maximum field length must be specified.

**Comment by flarco** [comment]
OK, after some research it seems that putting a very large default could consume more memory while loading (not 100% sure), so I'm leaving it at 400k. To override it, you need to specify the column type with a length (since this isn't a common situation).

```yaml
...
streams:
  my_table:
    columns:
      large_col: text(1000000)  # translates to CLOB type in oracle target
...
```

Can you test with latest [dev build?](https://github.com/slingdata-io/sling-cli?tab=readme-ov-file#automated-dev-builds)

**Comment by dvdrndlph** [comment]
No luck:
```
$ pyth cordite.py load --table ImagingTextFact --pk ImagingTextKey --chunk_size 10000 --debug --direct --big_clobs 'Narrative,Impression' --dev
fatal:
~ invalid columns -> 'Narrative:
yaml: found unexpected end of stream
The last pk value is 588
sling run --src-conn CDW_MSSQL --src-stream file://C:\cygwin64\tmp\ImagingTextFact_stream.sql --tgt-conn ORABOODLE --tgt-object ImagingTextFact --tgt-conn ORABOODLE --tgt-object ImagingTextFact --mode incremental --update-key ImagingTextKey --limit 10000 --columns 'Narrative: text(800000), Impression: text(800000)' -d
Iteration 1 failed.

drando2@UIVMSSISQ03 ~/repo/CCTS-ETL/accord
$ pyth cordite.py load --table ImagingTextFact --pk ImagingTextKey --chunk_size 10000 --debug --direct --big_clobs '"Narrative","Impression"' --dev
fatal:
~ invalid columns -> '"Narrative":
yaml: found unexpected end of stream
The last pk value is 588
sling run --src-conn CDW_MSSQL --src-stream file://C:\cygwin64\tmp\ImagingTextFact_stream.sql --tgt-conn ORABOODLE --tgt-object ImagingTextFact --tgt-conn ORABOODLE --tgt-object ImagingTextFact --mode incremental --update-key ImagingTextKey --limit 10000 --columns '"Narrative": text(800000), "Impression": text(800000)' -d
Iteration 1 failed.
```

**Comment by dvdrndlph** [comment]
No luck:
```
$ pyth cordite.py load --table ImagingTextFact --pk ImagingTextKey --chunk_size 10000 --debug --direct --big_clobs 'Narrative' --dev
fatal:
--- proc.go:271 main ---
--- sling_cli.go:452 main ---
--- sling_cli.go:488 cliInit ---
--- cli.go:286 CliProcess ---
~ invalid columns -> '{
--- sling_run.go:144 processRun ---
yaml: found unexpected end of stream
The last pk value is 588
sling run --src-conn CDW_MSSQL --src-stream file://C:\cygwin64\tmp\ImagingTextFact_stream.sql --tgt-conn ORABOODLE --tgt-object ImagingTextFact --tgt-conn ORABOODLE --tgt-object ImagingTextFact --mode incremental --update-key ImagingTextKey --limit 10000 --columns '{ Narrative: text(800000) }' -d
Iteration 1 failed.

drando2@UIVMSSISQ03 ~/repo/CCTS-ETL/accord
$ pyth cordite.py load --table ImagingTextFact --pk ImagingTextKey --chunk_size 10000 --debug --direct --big_clobs '"Narrative"' --dev
fatal:
~ invalid columns -> '{
yaml: found unexpected end of stream
The last pk value is 588
sling run --src-conn CDW_MSSQL --src-stream file://C:\cygwin64\tmp\ImagingTextFact_stream.sql --tgt-conn ORABOODLE --tgt-object ImagingTextFact --tgt-conn ORABOODLE --tgt-object ImagingTextFact --mode incremental --update-key ImagingTextKey --limit 10000 --columns '{ "Narrative": text(800000) }' -d
Iteration 1 failed.
```
What am I missing?

**Comment by flarco** [comment]
Are you using Powershell? if not, can you try? the quote/bracket can have weird behavior.

Just tested in linux, and worked fine for me (see ctrl file content below)

```shell
$ ./sling run --src-stream file://tests/files/test1.csv --tgt-conn oracle --tgt-object system.test1k -d --columns ' { FIRST_NAME: text(800000) } '
2024-11-24 00:48:22 DBG opened "file" connection (conn-file-TIg)
2024-11-24 00:48:23 DBG Sling version: dev (linux amd64)
2024-11-24 00:48:23 DBG type is file-db
2024-11-24 00:48:23 DBG using: {"columns":{"FIRST_NAME":"text(800000)"},"mode":"full-refresh","transforms":null}
2024-11-24 00:48:23 DBG using source options: {"trim_space":false,"empty_as_null":true,"header":true,"fields_per_rec":-1,"compression":"auto","null_if":"NULL","datetime_format":"AUTO","skip_blank_lines":false,"max_decimals":-1}
2024-11-24 00:48:23 DBG using target options: {"datetime_format":"auto","file_max_rows":0,"max_decimals":-1,"use_bulk":true,"add_new_columns":true,"adjust_column_type":false,"column_casing":"source"}
2024-11-24 00:48:23 DBG opened "oracle" connection (conn-oracle-cQZ)
2024-11-24 00:48:23 INF connecting to target database (oracle)
2024-11-24 00:48:23 INF reading from source file system (file)
2024-11-24 00:48:23 DBG opened "file" connection (conn-file-X5W)
2024-11-24 00:48:23 DBG reading single datastream from file://tests/files/test1.csv [format=csv]
2024-11-24 00:48:23 DBG merging csv readers of 1 files [concurrency=3] from file://tests/files/test1.csv
2024-11-24 00:48:23 DBG processing reader from file://tests/files/test1.csv
2024-11-24 00:48:23 DBG delimiter auto-detected: ","
2024-11-24 00:48:23 DBG casting column 'FIRST_NAME' as 'text'
2024-11-24 00:48:23 INF writing to target database [mode: full-refresh]
2024-11-24 00:48:23 DBG BEGIN
  EXECUTE IMMEDIATE 'DROP TABLE "SYSTEM"."TEST1K_TMP8J" purge';
EXCEPTION
  WHEN OTHERS THEN
    IF SQLCODE != -942 THEN
      RAISE;
    END IF;
END
2024-11-24 00:48:23 DBG table "SYSTEM"."TEST1K_TMP8J" dropped
2024-11-24 00:48:23 DBG BEGIN
  EXECUTE IMMEDIATE 'create table "SYSTEM"."TEST1K_TMP8J" ("ID" number(19),
"FIRST_NAME" clob,
"LAST_NAME" clob,
"EMAIL" clob,
"TARGET" varchar(5),
"CREATE_DT" timestamp(9),
"DATE" date,
"RATING" number(24,6),
"CODE" number(19),
"UPDATE_DT" timestamp(9),
"_SLING_LOADED_AT" number(10))';
EXCEPTION
  WHEN OTHERS THEN
    IF SQLCODE != -955 THEN
      RAISE;
    END IF;
END
2024-11-24 00:48:23 INF created table "SYSTEM"."TEST1K_TMP8J"
2024-11-24 00:48:23 INF streaming data
2024-11-24 00:48:23 DBG sqldr ctl content:
OPTIONS (
  SKIP=1,
  ROWS=1000,
  PARALLEL=true,
  DIRECT=true,
  SKIP_INDEX_MAINTENANCE=true
)
LOAD DATA
INFILE '/dev/stdin'
APPEND
INTO TABLE "SYSTEM"."TEST1K_TMP8J"
FIELDS TERMINATED BY ","
OPTIONALLY ENCLOSED BY '"'
TRAILING NULLCOLS
(
    "ID" ,
  "FIRST_NAME" char(800000) NULLIF "FIRST_NAME"=BLANKS,
  "LAST_NAME" char(400000) NULLIF "LAST_NAME"=BLANKS,
  "EMAIL" char(400000) NULLIF "EMAIL"=BLANKS,
  "TARGET" ,
  "CREATE_DT" "TO_TIMESTAMP(:\"CREATE_DT\", 'YYYY-MM-DD HH24:MI:SS.FF6')",
  "DATE" "TO_DATE(:\"DATE\", 'YYYY-MM-DD HH24:MI:SS')",
  "RATING" ,
  "CODE" ,
  "UPDATE_DT" "TO_TIMESTAMP(:\"UPDATE_DT\", 'YYYY-MM-DD HH24:MI:SS.FF6')",
  "_SLING_LOADED_AT" 
)

2024-11-24 00:48:23 DBG sqlldr '****' control=/tmp/oracle.system.test1k_tmp8j.sqlldr.1732409303633.esK.ctl discardmax=0 errors=0 data=/dev/stdin log=/dev/stdout bad=/dev/stdout
2024-11-24 00:48:25 DBG column type change for code (bigint to decimal)
2024-11-24 00:48:25 DBG UPDATE "SYSTEM"."TEST1K_TMP8J" SET "FIRST_NAME" = REPLACE(REPLACE("FIRST_NAME", chr(13)), '~/N/~', chr(10))
2024-11-24 00:48:25 DBG select count(*) cnt from "SYSTEM"."TEST1K_TMP8J"
2024-11-24 00:48:25 DBG BEGIN
  EXECUTE IMMEDIATE 'DROP TABLE "SYSTEM"."TEST1K" purge';
EXCEPTION
  WHEN OTHERS THEN
    IF SQLCODE != -942 THEN
      RAISE;
    END IF;
END
2024-11-24 00:48:26 DBG table "SYSTEM"."TEST1K" dropped
2024-11-24 00:48:26 DBG BEGIN
  EXECUTE IMMEDIATE 'create table "SYSTEM"."TEST1K" ("ID" number(10),
"FIRST_NAME" clob,
"LAST_NAME" clob,
"EMAIL" clob,
"TARGET" varchar(5),
"CREATE_DT" timestamp(9),
"DATE" date,
"RATING" number(24,6),
"CODE" number(24,6),
"UPDATE_DT" timestamp(9),
"_SLING_LOADED_AT" number(19))';
EXCEPTION
  WHEN OTHERS THEN
    IF SQLCODE != -955 THEN
      RAISE;
    END IF;
END
2024-11-24 00:48:26 INF created table "SYSTEM"."TEST1K"
2024-11-24 00:48:26 DBG insert into "SYSTEM"."TEST1K" ("ID", "FIRST_NAME", "LAST_NAME", "EMAIL", "TARGET", "CREATE_DT", "DATE", "RATING", "CODE", "UPDATE_DT", "_SLING_LOADED_AT") select "ID", "FIRST_NAME", "LAST_NAME", "EMAIL", "TARGET", "CREATE_DT", "DATE", "RATING", "CODE", "UPDATE_DT", "_SLING_LOADED_AT" from "SYSTEM"."TEST1K_TMP8J"
2024-11-24 00:48:26 DBG inserted rows into "SYSTEM"."TEST1K" from temp table "SYSTEM"."TEST1K_TMP8J"
2024-11-24 00:48:26 INF inserted 1000 rows into "SYSTEM"."TEST1K" in 3 secs [308 r/s]
2024-11-24 00:48:26 DBG BEGIN
  EXECUTE IMMEDIATE 'DROP TABLE "SYSTEM"."TEST1K_TMP8J" purge';
EXCEPTION
  WHEN OTHERS THEN
    IF SQLCODE != -942 THEN
      RAISE;
    END IF;
END
2024-11-24 00:48:27 DBG table "SYSTEM"."TEST1K_TMP8J" dropped
2024-11-24 00:48:27 DBG closed "oracle" connection (conn-oracle-cQZ)
2024-11-24 00:48:27 INF execution succeeded
```

**Comment by dvdrndlph** [comment]
I am making a system call from a Python script like so:
```
result = subprocess.run(cmd.split(), capture_output=False, text=True)
```
where the command is just the string I would execute in PowerShell. The split is separating by whitespace, so maybe that is the problem. None of the other tokens have any spaces.

**Comment by dvdrndlph** [comment]
Okay, I needed to get rid of the split and double-quote the argument to `--columns`. The following works:
```
sling run --src-conn CDW_MSSQL --src-stream file://C:\cygwin64\tmp\ImagingTextFact_stream.sql --tgt-conn ORABOODLE --tgt-object ImagingTextFact --tgt-conn ORABOODLE --tgt-object ImagingTextFact --mode incremental --update-key ImagingTextKey --limit 200 --columns "{ Narrative: text(800000) }"  -d
```

**Comment by dvdrndlph** [comment]
Hang on. This runs, but it does not modify the .ctl file accordingly.

**Comment by dvdrndlph** [comment]
It works with either kind of quotation marks in PowerShell. So there is something going on with the system call in Python. I am a bit baffled by this. Is it time to explore your Python wrapper?

**Comment by dvdrndlph** [comment]
Okay, I removed all the quotation marks loaded up a list, and it works. Here is a Python snippet:
```
columns_arg = "{ big_field: text(800000) }" # We quote nothing!
cmd = f"sling run --src-conn {source_conn} --src-stream file://{stream_file_path} " +\
      f"--tgt-conn {target_conn} --tgt-object {table} " +\
      f"--mode incremental --update-key {pk} " +\
      f"--limit {chunk_size} -d "

cmd_tokens = cmd.split()
if columns_arg:
    cmd_tokens.extend(["--columns", columns_arg])

print(" ".join(cmd_tokens))
result = subprocess.run(cmd_tokens, capture_output=False, text=True)
```
